File: attribouter/src/main/java/me/jfenn/attribouter/wedges/Wedge.java
package me.jfenn.attribouter.wedges;

import android.content.Context;
import android.content.res.XmlResourceParser;
import androidx.annotation.LayoutRes;
import androidx.recyclerview.widget.RecyclerView;
import android.util.Log;
import android.view.View;

import org.xmlpull.v1.XmlPullParserException;

import java.io.IOException;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.List;

import me.jfenn.attribouter.data.github.GitHubData;
import me.jfenn.attribouter.interfaces.Mergeable;

public abstract class Wedge<T extends Wedge.ViewHolder> implements GitHubData.OnInitListener {

    private int layoutRes;
    private List<Wedge> children;
    private List<GitHubData> requests;

    private OnRequestListener listener;

    public Wedge(@LayoutRes int layoutRes) {
        this.layoutRes = layoutRes;
        requests = new ArrayList<>();
        children = new ArrayList<>();
    }

    protected void addChildren(XmlResourceParser parser) throws IOException, XmlPullParserException {
        while (parser.next() != XmlResourceParser.END_TAG || !parser.getName().equals(getClass().getName())) {
            if (parser.getEventType() == XmlResourceParser.START_TAG) {
                try {
                    Class<?> classy = Class.forName(parser.getName());
                    Constructor constructor = classy.getConstructor(XmlResourceParser.class);
                    addChild((Wedge) constructor.newInstance(parser));
                } catch (ClassNotFoundException e) {
                    Log.e("Attribouter", "Class name \"" + parser.getName() + "\" not found - you should probably check your configuration file for typos.");
                    e.printStackTrace();
                } catch (NoSuchMethodException e) {
                    Log.e("Attribouter", "Class \"" + parser.getName() + "\" definitely exists, but doesn't have the correct constructor. Check that you have defined one with a single argument - \'android.content.res.XmlResourceParser\'");
                    e.printStackTrace();
                } catch (IllegalAccessException e) {
                    e.printStackTrace();
                } catch (java.lang.InstantiationException e) {
                    e.printStackTrace();
                } catch (InvocationTargetException e) {
                    e.printStackTrace();
                } catch (ClassCastException e) {
                    Log.e("Attribouter", "Class \"" + parser.getName() + "\" has been instantiated correctly, but it must extend \'me.jfenn.attribouter.data.info.InfoData\' to be worthy of the great RecyclerView adapter.");
                    e.printStackTrace();
                }
            }
        }
    }

    Wedge addChild(Wedge child) {
        return addChild(children.size(), child);
    }

    Wedge addChild(int index, Wedge child) {
        if (!children.contains(child)) {
            children.add(index, child);
        } else {
            Wedge merger = children.get(children.indexOf(child));
            if (merger instanceof Mergeable) {
                ((Mergeable) merger).merge(child);
                return merger;
            } else {
                children.add(index, child);
            }
        }

        if (listener != null)
            child.setOnRequestListener(listener);

        return child;
    }

    public List<Wedge> getChildren() {
        return children;
    }

    public <X extends Wedge> List<X> getChildren(Class<X> type) {
        List<X> children = new ArrayList<>();
        for (Wedge info : getChildren()) {
            if (type.isInstance(info))
                children.add((X) info);
        }

        return children;
    }

    protected void addRequest(GitHubData request) {
        request.addOnInitListener(this);
        requests.add(request);
        if (listener != null)
            listener.onRequest(this, request);
    }

    public final void setOnRequestListener(OnRequestListener listener) {
        this.listener = listener;
        if (listener != null) {
            for (GitHubData request : requests)
                listener.onRequest(this, request);
        }

        for (Wedge child : children)
            child.setOnRequestListener(listener);
    }

    public final List<GitHubData> getRequests() {
        return requests;
    }

    public final boolean hasRequest(GitHubData request) {
        return requests.contains(request);
    }

    public final int getLayoutRes() {
        return layoutRes;
    }

    public abstract T getViewHolder(View v);

    public abstract void bind(Context context, T viewHolder);

    @Override
    public void onInit(GitHubData data) {
    }

    @Override
    public void onFailure(GitHubData data) {
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {

        public ViewHolder(View v) {
            super(v);
        }

    }

    public interface OnRequestListener {
        void onRequest(Wedge info, GitHubData request);
    }

}

File: attribouter/src/main/java/me/jfenn/attribouter/wedges/Wedge.kt
package me.jfenn.attribouter.wedges

import android.content.Context
import android.view.View
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.RecyclerView
import me.jfenn.attribouter.interfaces.Notifiable
import me.jfenn.attribouter.provider.LifecycleInstance
import me.jfenn.attribouter.provider.wedge.WedgeProvider
import me.jfenn.attribouter.utils.isResourceMutable
import kotlin.reflect.KProperty

abstract class Wedge<T : Wedge.ViewHolder>(@param:LayoutRes val layoutRes: Int) : Notifiable {

    private val children: MutableList<Wedge<*>> = ArrayList()
    private val attributes: ArrayList<attr<*,*>> = ArrayList()
    internal var lifecycle: LifecycleInstance? = null
    internal var parent: Notifiable? = null

    open var isHidden by object : attr<Wedge<T>, Boolean>("hidden", false) {
        override fun apply(original: Boolean?, value: Boolean?): Boolean? = original == true || value == true
    }

    fun <R: Wedge<*>> create(lifecycle: LifecycleInstance? = null) : R {
        this.lifecycle = lifecycle

        onCreate()
        return this as R
    }

    /**
     * Combine one Wedge's properties with another's...
     *
     * Used to handle duplicate links or contributors, based on `.equals()`
     */
    fun merge(other: Wedge<*>) : Wedge<*> {
        for (otherAttr in other.attributes) {
            attributes.find { it.name == otherAttr.name }?.mergeValue(otherAttr.getValue(null, null))
        }

        notifyItemChanged()
        return this
    }

    fun withWedgeProvider(provider: WedgeProvider) : Wedge<*> {
        for (attribute in attributes)
            attribute.withProvider(provider)

        addChildren(provider.getWedges(this))
        return this
    }

    fun withParent(parent: Notifiable) : Wedge<*> {
        this.parent = parent
        return this
    }

    /**
     * Handle changes of child items.
     *
     * By default, simply forwards events to the next parent,
     * causing the entire view to be re-bound.
     */
    override fun onItemChanged(changed: Wedge<*>) {
        notifyItemChanged()
    }

    open fun onCreate() {}

    /**
     * Shortcut function for LifecycleInstance
     */
    internal fun notifyItemChanged() {
        (parent ?: lifecycle?.notifiable)?.onItemChanged(this)
    }

    internal fun addChildren(children: Collection<Wedge<*>>) {
        for (c in children) addChild(c)
    }

    internal fun addChild(child: Wedge<*>): Wedge<*> {
        return addChild(children.size, child)
    }

    internal fun addChild(index: Int, child: Wedge<*>): Wedge<*> {
        if (!children.contains(child)) {
            children.add(index, child.withParent(this))
        } else {
            children[children.indexOf(child)].merge(child)
        }

        return child
    }

    fun getChildren(): List<Wedge<*>> {
        return children
    }

    inline fun <reified X : Wedge<*>> getTypedChildren(): List<X> {
        return getChildren().filterIsInstance<X>()
    }

    abstract fun getViewHolder(v: View): T

    abstract fun bind(context: Context, viewHolder: T)

    open class ViewHolder(v: View) : RecyclerView.ViewHolder(v)

    open inner class attr<in R : Wedge<*>, T>(
            val name: String,
            protected var property: T? = null
    ) {

        init {
            attributes.add(this)
        }

        open fun withProvider(provider: WedgeProvider) {
            property = provider.getAttribute(this@Wedge, name, property) ?: property
        }

        open fun apply(original: T?, value: T?) : T? {
            return value ?: original
        }

        operator fun getValue(thisRef: R?, prop: KProperty<*>?): T {
            return if ((property as? String)?.startsWith('^') == true)
                (property as? String)?.substring(1) as T
            else property as T
        }

        open operator fun setValue(thisRef: R?, prop: KProperty<*>?, value: T?) {
            if ((property as? String).isResourceMutable())
                property = apply(property, value)
        }

        open fun mergeValue(value: Any?) {
            (value as? T?).let { setValue(null, null, it) }
        }
    }

    open inner class attrInt<in R: Wedge<*>>(
            attribute: String,
            property: Int? = null
    ) : attr<R, Int>(attribute, property) {

        override fun withProvider(provider: WedgeProvider) {
            provider.getAttribute(this@Wedge, name, property.toString())?.let {
                property = Integer.parseInt(it)
            }
        }

    }

}
